Distributed execution via worker

michelvocks opened the pull request:
Fixes #107 and #46 

Related docs PR: https://github.com/gaia-pipeline/docs/pull/11

TODO:
- [x] Add tags to worker UI
- [x] Fix tab panel order in UI
- [x] Add option to prevent primary instance to accept work
- [x] Write more tests
- [x] Write documentation

michelvocks committed:
First tests

michelvocks committed:
A bit more work on the register api

michelvocks committed:
Implemented registration process

michelvocks committed:
little work

michelvocks committed:
Removed trash from last merge conflict

michelvocks committed:
Growing

michelvocks committed:
Implemented load of client tls certificates

michelvocks committed:
Some more work

michelvocks committed:
Added new parameters

michelvocks committed:
Tested worker registration process and fixed several bugs

michelvocks committed:
WIP server grpc implementation

michelvocks committed:
WIP memdb implementation

michelvocks committed:
More work on memdb implementation

michelvocks committed:
Added memdb to modules

michelvocks committed:
More work done on getwork

michelvocks committed:
Implemented binary streaming for pipeline binaries

michelvocks committed:
More work done

michelvocks committed:
Work work work

michelvocks committed:
Implemented updatework function

michelvocks committed:
Finalizing for first test

michelvocks committed:
More work done

michelvocks committed:
Massive work done. Refactored tons of stuff. Pray that it will not break something :-O

michelvocks committed:
Implemented logs shipping

michelvocks committed:
More work done and some testing

michelvocks committed:
IT WORKS LOL

michelvocks committed:
More bug fixing

michelvocks committed:
Implemented authentication and deregister function

Skarlso commented:
Duuddee! Nice, I'm going to have to review this at one point. :D

Skarlso commented:
BTW, this should have been a bit separated into multiple PRs if possible. :D Like, the worker view, then the pipeline stuff and so on and so forth. :)

michelvocks commented:
@Skarlso I thought about splitting this PR up in multiple PRs beforehand but this almost impossible. The way this has been implemented now is a bit different from what we agreed on in the RFC since some things didn't work well or included some traps we didn't think of. 😭 

Let me check if I can split this up when it's finished. 😄

Skarlso commented:
@michelvocks It's fine, it will just be harder to review. :D 

Ah, would you mind updating the RFC in that case? :P

michelvocks committed:
Implemented backend worker tags

michelvocks committed:
More work on the frontend

michelvocks committed:
Implemented status view of workers. Added worker permissions. Implemented automatic primary instance work scheduling when workers are busy.

michelvocks committed:
Added worker data fetch

michelvocks committed:
Implemented deregister process and some bug fixing

michelvocks committed:
Implemented reset of worker secret. Implemented ignored specialized vault keys

michelvocks committed:
Implemented pipeline worker tags. WIP backend testing

michelvocks committed:
Implemented automatic worker status transission and a few more status checks

michelvocks committed:
Added worker registration test and fixed a few tests

michelvocks committed:
Fixed most tests

michelvocks committed:
Upgraded circleci

michelvocks committed:
Fixed last failing tests

michelvocks committed:
circle ci

michelvocks committed:
Fixed data race

michelvocks committed:
fixed

michelvocks committed:
fixed

michelvocks committed:
Linter fixes

michelvocks committed:
fix

michelvocks committed:
fix tests

michelvocks committed:
Fixed

michelvocks committed:
Fixed

michelvocks committed:
Added more workers handler tests

michelvocks committed:
Fixed concurrent tests

michelvocks committed:
Added memdb service test

michelvocks commented:
@Skarlso This is ready for review. Need to work on a few tests and documentation but overall ready for initial review.

Skarlso commented:
wooot!! Holy Moly! :D I'll test it out... The code review is massive. :D So that will take some time. :P

michelvocks committed:
Fixed failing test

Skarlso commented:
`+5,700 −954`

Nooooooo. :D

michelvocks commented:
@Skarlso Whoopsi 😄

Skarlso commented on handlers/handler.go:
```diff
@@ -82,12 +85,20 @@ func InitHandlers(e *echo.Echo) error {
 
+	// Worker
+	e.GET(p+"worker/secret", GetWorkerRegisterSecret)
+	e.POST(p+"worker/register", RegisterWorker)
```
To follow some practices, this could be just simply `worker` and a `POST` since GET `worker` gets a worker. I don't think you need the `register` part.

michelvocks replied:
Good catch. Thanks!

Skarlso commented on helper/stringhelper/stringhelper.go:
```diff
@@ -0,0 +1,16 @@
+
+// IsContainedInSlice checks if the given string is contained in the slice.
+// If case insensitive is enabled, checks are case insensitive.
+func IsContainedInSlice(s []string, c string, caseIns bool) bool {
```
This is never called with caseIns = false. :)

michelvocks replied:
Not yet. I just want to make sure that its there when someone needs it 😃

Skarlso commented on handlers/pipeline_test.go:
```diff
@@ -28,7 +28,7 @@ type mockScheduleService struct {
 }
 
-func (ms *mockScheduleService) SchedulePipeline(p *gaia.Pipeline, args []gaia.Argument) (*gaia.PipelineRun, error) {
+func (ms *mockScheduleService) SchedulePipeline(p *gaia.Pipeline, args []*gaia.Argument) (*gaia.PipelineRun, error) {
```
How come this is now a pointer?

michelvocks replied:
I found some inconsistencies when and where we are using pointer or values. I just changed it now to be consistent (usually pointers).

Skarlso commented:
so this will take a little while... I want to go over it on a high level.. and then start to go over it structurally and then architecturally.... :D But I'll try to hurry things up so it doesn't age too much.

Probably will be fine at first then as we go we work the kinks out.

michelvocks commented:
> so this will take a little while... I want to go over it on a high level.. and then start to go over it structurally and then architecturally.... :D But I'll try to hurry things up so it doesn't age too much.
> 
> Probably will be fine at first then as we go we work the kinks out.

Yep, no problem. A lot of tests are still missing and the documentation.

Skarlso commented on workers/agent/agent.go:
```diff
@@ -0,0 +1,686 @@
+	}
+
+	// Set path to local certificates
+	ag.certFile = filepath.Join(gaia.Cfg.HomePath, "cert.pem")
```
Shouldn't this be a separate configuration entity since agents can be in a different location then gaia Home?

michelvocks replied:
Agents use `HomePath` to store their local data which must be different from the primary home directory. I just thought there is no reason to create a new path variable for the agents.

Skarlso commented on workers/agent/agent.go:
```diff
@@ -0,0 +1,686 @@
+	// Send all pipeline runs to the remote primary instance
+	for _, run := range runs {
+		// Transform to protobuf struct
+		runPB := &pb.PipelineRun{
```
So, bare with me on this one. What I started to do is instead of creating a struct and assigning the address of it to a variable, I started to do the address of it at the point I want to share it. To clarify, the above code would be like this:

```go
			j := pb.Job{
				UniqueId:    job.ID,
				Title:       job.Title,
				Status:      string(job.Status),
				Description: job.Description,
			}
			runPB.Jobs = append(runPB.Jobs, &j)
```

It's a very subtle change, but makes it clear that at THAT point you are sharing this struct. What say you?

michelvocks replied:
I'm wondering if there is some kind of convention? Happy to stick to one solution

Skarlso replied:
Not that I know of. I think it's just usage habit. Whatever suits you better I guess. :)

Skarlso commented on workers/agent/agent.go:
```diff
@@ -0,0 +1,686 @@
+
+// updateWork is function that is periodically called and it is used to
+// send new information about a pipeline run to the Gaia master instance.
+func (a *Agent) updateWork() {
```
This and scheduleWork REEALLLY need to be broken down. These functions are way too large man. :)

michelvocks replied:
Yeah. I still need to write tests for them anyway so I think I have to refactor them a bit.

Skarlso commented:
Uh, something's not right. The tables are completely out of my window proportions.
<img width="1440" alt="Screenshot 2019-06-04 at 22 29 01" src="https://user-images.githubusercontent.com/182850/58911514-4221ff80-8718-11e9-8108-7a5b6dfff7ad.png">

Skarlso commented:
Huh. So I tried creating a pipeline. It said it created it successfully, but it's nowhere to be seen. :D 
So Gaia master can no longer deal with pipelines?

michelvocks committed:
Worked on tests

michelvocks committed:
Added missing error check

michelvocks commented:
@Skarlso I guess you need a bigger screen 😄 
The problem was that the table panel was too big. I renamed the panels and removed "Manage" from them.

michelvocks committed:
Fixed racy test

Skarlso commented:
Alright, looks good now. :)

Skarlso commented:
@michelvocks how do you register a worker? :D

Skarlso commented:
Where is the agent code? I thought there will be a gaia systemd daemon somewhere with which Gaia can communicate?

But it looks like Gaia itself should be running on the agent?

Skarlso commented:
So, I'm trying to register a worker. And I'm getting this:
```
wrong global worker secret provided
```

This means that the secret that is provided in the menu called Worker registration secret: did not equal with that secret. But I can't change it. So.... what's up? :D

Skarlso commented:
NO idea why this is not working:
`-d '{"secret":"55695ea0-f75e-5b99-b4a3-429c3e3f1b9b","name":"test_worker","tags":["tag1"]}'`

Skarlso commented on handlers/vault.go:
```diff
@@ -37,6 +39,12 @@ func SetSecret(c echo.Context) error {
 		key = s.Key
 		value = s.Value
 	}
+
```
It looks like this has been already put in here however, not:

```go

type addSecret struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type updateSecret struct {
	Key   string `json:"key"`
	Value string `json:"newvalue"`
}
```

You don't need to do this. Json accepts multiple values:

```go
type updateSecret struct {
	Key   string `json:"key"`
	Value string `json:"value,newvalue"`
}
```

This way you would only have one struct. :)

michelvocks replied:
You did this. I thought the same when I saw that 😄

Skarlso replied:
OH Gods. What was I thinking? :D :D :D :D

Skarlso commented:
It's getting late. :) Some documentation is necessary dude. :) So I received back the cert, but where do I put it? :D I need to browse the code for that. :) I'm guessing that's why the worker is inactive?

A few suggestions though. I would fail the pipeline if the tag is not found. Saying no matching machines with tag `bla` instead of just running it on master.

Also, display the tags for a worker. Right now I can't see what tags my worker has. :)

With that, good night!

Skarlso commented:
Okay, I promise I will stop writing on this PR because it's getting huge. :D And I always write all my nuances here. I'll open a Slack channel for debugging / testing this PR and if there is a bug or something I'll comment that. Otherwise this will be a very long PR and hard to follow what's me rambling and an actual bug. :)

michelvocks commented:
> how do you register a worker? :D

The agent is built into the normal binary. Makes it easier to distribute the binary (only one binary for everything). To register a worker you need to set a few parameters. This is the config I used for testing:
```
hosturl=http://localhost:8080
mode=worker
workername=my-worker
grpchosturl=localhost:8989
workersecret=cfd7b91c-ae5b-5872-9e1f-28592a935ee2
```

> Where is the agent code? I thought there will be a gaia systemd daemon somewhere with which Gaia can communicate?
> 
> But it looks like Gaia itself should be running on the agent?

When you start Gaia, you define with the `mode` parameter in which mode Gaia should start (worker or server).

> It's getting late. :) Some documentation is necessary dude. :) So I received back the cert, but where do I put it? :D I need to browse the code for that. :) I'm guessing that's why the worker is inactive?

Yeah don't worry. Documentation will follow. Just wanted to finish the tests first. :-)
When the worker starts it will automatically look for an existing certificate. If the certificate is not there it will try to register at the primary instance which will attach the mTLS certs in response and place them inside your home folder. So by default, you should do nothing with it since everything should work out of the box. Are there any errors in the logs or something else?

Here is a brief list of steps you need to do that it works:

1. Start Gaia as usual.
2. Optional: Compile the gaia binary and place it in a different folder.
3. Create a config file `.gaia_config` like above (don't forget to replace the secret) in the folder with the binary.
4. Start the worker with the following command: `./gaia -homepath=${PWD}/tmp`

> A few suggestions though. I would fail the pipeline if the tag is not found. Saying no matching machines with tag `bla` instead of just running it on master.

My idea is to add an option which prevents Gaia to run pipelines on the primary instance.

> Also, display the tags for a worker. Right now I can't see what tags my worker has. :)

Good point. Will add that.

michelvocks committed:
More work on tests

michelvocks commented:
@Skarlso Okay, I know why the registration process was not working. I changed the registration API endpoint and then it was secured (you needed to be registered). I fixed that by changing the registration path back to `worker/registration`. This is the only way to make this path "insecure" without affecting the other endpoints.

Skarlso commented:
@michelvocks  Oh I see. That's why it was there. :D Okay.

michelvocks committed:
fixed wrong fixtures path

Skarlso commented:
@michelvocks BTW. Instead of the horrible, growing endpoint if exclusion thing we have going on, why aren't we using echo's grouping feature? https://echo.labstack.com/guide/routing

michelvocks commented:
@Skarlso I'm not really getting how the group feature would help us here? As far as I understand the grouping feature, you still have to define every endpoint.

michelvocks committed:
Added more test

Skarlso commented:
@michelvocks instead of the if, everything that's in the admin group would require authentication to access. Everything outside is public. No more if.

michelvocks committed:
Added server worker tests

Skarlso commented on helper/filehelper/filehelper_test.go:
```diff
@@ -0,0 +1,32 @@
+	tmp, _ := ioutil.TempDir("", "TestGetSHA256Sum")
+	sumText := []byte("hello world\n")
+	filePath := filepath.Join(tmp, "test.file")
+	err := ioutil.WriteFile(filePath, sumText, 777)
```
I love this one!

michelvocks committed:
Fixed small test issues

michelvocks committed:
More work on tests

michelvocks committed:
Fixed small issues

michelvocks committed:
Increased waiting time

michelvocks committed:
Racy test fix

michelvocks committed:
growing

michelvocks committed:
fix

michelvocks committed:
fix

michelvocks committed:
Added missing piece

michelvocks committed:
Added worker tags to UI. Added option to prevent work being scheduled on primary.

michelvocks committed:
Improved test coverage

michelvocks commented:
@Skarlso Tests are looking fine now. The test coverage will be decreased a bit but mainly because codecov sucks 😞 I touched a few files where we didn't have any tests and codecov simply counts them now as new code 🙄 

The plan is to finish the documentation part within the next week. Would be great if you could have another look until then. 😄

Skarlso commented:
@Skarlso sure thing dude. :) It's kind of hard to review 6000 lines of code change though... :D

Skarlso commented:
Noice!
<img width="1182" alt="Screenshot 2019-06-16 at 15 06 46" src="https://user-images.githubusercontent.com/182850/59564549-63ec8200-9048-11e9-808c-c9f0444c6daf.png">

Skarlso commented:
@michelvocks Is there a was to delete the default tags? I mean what if my worker isn't supporting the Python environment? :)

Skarlso commented:
@michelvocks  Also, the pipelines need a way to display their affiliated tags. I can't view what environment my pipeline needs once I set a tag.

michelvocks committed:
Added negative tags to allow users to remove automatically detected tags. Added tags info to overview view. Fixed add tags bug.

michelvocks committed:
Removed prettify cpp tag since it might lead to confusion

michelvocks commented:
> @Skarlso sure thing dude. :) It's kind of hard to review 7700 lines of code change though... :D

I don't expect you to review all the lines of code. 😄 
It already helps a lot if you just have a general look and may provide feedback like above. Super helpful 🤗 

> @michelvocks Is there a was to delete the default tags? I mean what if my worker isn't supporting the Python environment? :)

I've added the possibility to provide "negative tags". So if you provide a tag like "-golang" it will remove automatically the "golang" tag (and the "-golang" tag). Not the best solution in terms of user experience but should work for this use-case.  😄 

> @michelvocks Also, the pipelines need a way to display their affiliated tags. I can't view what environment my pipeline needs once I set a tag. I thought that maybe we'd need the ability to edit tags, but I reconsidered. In case an environment need changes for a pipeline we'd just delete it and re-create it. But we still need an ability to see them tags. :)

Whops, yeah. I added the tags to the overview page.

Skarlso commented:
@michelvocks awesome, I'm okay with the negative tags. :D so do you tag a pipeline automatically when it's created with the appropriate environment tag? Like a Go pipeline gets go a Python pipeline gets Python etc?

michelvocks commented:
> @michelvocks awesome, I'm okay with the negative tags. :D so do you tag a pipeline automatically when it's created with the appropriate environment tag? Like a Go pipeline gets go a Python pipeline gets Python etc?

Yep: https://github.com/gaia-pipeline/gaia/pull/166/files#diff-f88a32f252fb4994c4b7249afcc79b6cR57

michelvocks commented:
@Skarlso Docs PR: https://github.com/gaia-pipeline/docs/pull/11

Still WIP. Need to update the RFC.

michelvocks commented:
@Skarlso Docs PR should be fine now. In my opinion, this is now ready for merge.

Let me know what you think. 🤗

Skarlso commented on workers/agent/agent.go:
```diff
@@ -0,0 +1,713 @@
+	// Evaluate and summarize all worker tags
+	var tags []string
+	if gaia.Cfg.WorkerTags != "" {
+		trimmedTags := strings.ReplaceAll(gaia.Cfg.WorkerTags, " ", "")
```
@michelvocks hm, this means that we aren't allowing for tags with spaces in them. That is okay, I guess, right?

michelvocks replied:
Correct. Tags are single words which can be combined with dashes (for example). From my experience, that is the expected behavior.

Skarlso commented:
@michelvocks Well. Functionality wise this looks okay. I tested it and it worked. :D But man, the code needs refactoring. :DDD Create a bunch of follow up tickets so we can track the most common ones? Like the functions are too long and such? Otherwise, go ahead. :)

michelvocks commented:
> @michelvocks Well. Functionality wise this looks okay. I tested it and it worked. :D But man, the code needs refactoring. :DDD Create a bunch of follow up tickets so we can track the most common ones? Like the functions are too long and such? Otherwise, go ahead. :)

Is it only the length of some functions? In my opinion, it is quite common and normal to have a bit longer functions. It makes it often more readable because you can simply follow one block of code instead of switching between several functions. I also know other big projects (Kubernetes, Vault, Terraform etc.) that bigger functions can be beneficial.

michelvocks commented:
ping @Skarlso

Skarlso commented:
I disagree with that. That is why cyclomatic complexity is a thing. And that is why lingers are saying that your functions is too long. It's harder to track down a bug in a big ass function than in a small function where it's basically clear at one quick look what it does. If I have scroll down several pages and still be in the same function I'll loose track of what the beginning of the function was even trying to do in the first place. 😂 Also if a function is huge it's more likely that it does not adhere to SRP. The single responsibility principle. It is likely however that its trying to do everything / too much of its own.

I approve but we have to refractor these into easily tracakble understandable fragments I think. 
😊

Also, sorry for the late response, I'm on vacation in Austria. 😁
